src/job.ts

```diff
@@ -181,6 +182,26 @@ export class Job {
         predefinedVariables["CI_REGISTRY"] = `local-registry.${this.gitData.remote.host}`;
         predefinedVariables["CI_REGISTRY_IMAGE"] = `$CI_REGISTRY/${this._variables["CI_PROJECT_PATH"].toLowerCase()}`;
 
+        // Expand variables in rules:changes
+        if (this.rules && expandVariables) {
+            const expanded = Utils.expandVariables(this._variables);
+            this.rules.forEach((rule, ruleIdx) => {
+                const changes = Array.isArray(rule.changes) ? rule.changes : rule.changes?.paths;
+                if (!changes || !Array.isArray(changes)) {
+                    return;
+                }
+
+                changes.forEach((change, changeIdx) => {
+                    changes[changeIdx] = Utils.expandText(change, expanded);
+                });
+
+                assert(this.rules);
+                assert(this.rules[ruleIdx]);
```

Comment:
I'm not sure why the typescript compiler thinks `this.rules` or `this.rules[ruleIdx]` can be undefined at this point, but adding the asserts is required for the assignment below.